streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -1165,11 +1151,21 @@ private Thread shutdownHelper(final boolean error) {
             metrics.close();
             if (!error) {
                 setState(State.NOT_RUNNING);
+            } else {
+                setState(State.ERROR);
             }
         }, "kafka-streams-close-thread");
     }
 
     private boolean close(final long timeoutMs) {
+        if (state == State.ERROR) {
+            log.info("Streams client is already in the terminal state ERROR, all resources are closed and the client has stopped.");
+            return false;
```

Comment:
Should we return `true` there? I understand that we are not in `NOT_RUNNING` state, but in the end we are in a terminal state and we did cleanup all resources. -- I guess, I am raising the question if we should relax the definition of this return value?

Reply:
Fair point. Might be good to get input from somebody else in addition. \cc @cadonna

Reply:
As far as I can see, where the return value is used the javadoc says

```
true if all threads were successfully stopped, false if the timeout was reached.
```

Since all threads were successfully stopped, I would return `true`. We clearly document that `ERROR` is a terminal state, so I do not see why somebody should wait for `NOT_RUNNING` when the client is in `ERROR` and `close()` returns `true`.

Reply:
Alright I adjusted the close response to align with this.